flake: Router test "address already in use"

https://storage.googleapis.com/origin-ci-test/logs/test_branch_origin_end_to_end/424/build-log.txt

```
=== RUN   TestRouter
--- PASS: TestRouter (11.87s)
	router_test.go:353: TC non-secure: url 0.0.0.0 alias unsecure.example.com protocol http
	router_test.go:353: TC non-secure-path: url 0.0.0.0/test alias unsecure.example.com protocol http
	router_test.go:353: TC preferred-port: url 0.0.0.0 alias pref.port.unsecure.example.com protocol http
	router_test.go:353: TC edge termination: url 0.0.0.0 alias edge.example.com protocol https
	router_test.go:353: TC edge termination path: url 0.0.0.0/test alias edge.example.com protocol https
	router_test.go:353: TC reencrypt: url 0.0.0.0 alias reencrypt.example.com protocol https
	router_test.go:353: TC reencrypt-InsecureEdgePolicy: url 0.0.0.0 alias allow.reencrypt.example.com protocol http
	router_test.go:353: TC reencrypt-destcacert: url 0.0.0.0 alias destcert.reencrypt.example.com protocol https
	router_test.go:353: TC reencrypt path: url 0.0.0.0/test alias reencrypt.example.com protocol https
	router_test.go:353: TC passthrough termination: url 0.0.0.0 alias passthrough.example.com protocol https
	router_test.go:353: TC websocket unsecure: url 0.0.0.0 alias ws.unsecure.example.com protocol ws
	router_test.go:353: TC ws edge termination: url 0.0.0.0 alias ws.edge.example.com protocol wss
	router_test.go:353: TC ws passthrough termination: url 0.0.0.0 alias ws.passthrough.example.com protocol wss
=== RUN   TestRouterPathSpecificity
--- FAIL: TestRouterPathSpecificity (0.00s)
	router_test.go:393: Unable to start http server: listen tcp 172.18.1.71:8080: bind: address already in use
=== RUN   TestRouterDuplications
--- FAIL: TestRouterDuplications (0.00s)
	router_test.go:752: Unable to start http server: listen tcp 172.18.1.71:8080: bind: address already in use
=== RUN   TestRouterStatsPort
--- FAIL: TestRouterStatsPort (0.00s)
	router_test.go:884: Unable to start http server: listen tcp 172.18.1.71:8080: bind: address already in use
=== RUN   TestRouterHealthzEndpoint
--- FAIL: TestRouterHealthzEndpoint (0.00s)
	router_test.go:937: Unable to start http server: listen tcp 172.18.1.71:8080: bind: address already in use
=== RUN   TestRouterServiceUnavailable
--- FAIL: TestRouterServiceUnavailable (0.00s)
	router_test.go:975: Unable to start http server: listen tcp 172.18.1.71:8080: bind: address already in use
=== RUN   TestRouterReloadCoalesce
--- PASS: TestRouterReloadCoalesce (15.78s)
=== RUN   TestRouterBindsPortsAfterSync
--- PASS: TestRouterBindsPortsAfterSync (2.02s)
=== RUN   TestIngressConfiguredRouter
--- PASS: TestIngressConfiguredRouter (3.95s)
FAIL
```

It *looks* like the first router test didn't successfully shut something down (like the test server that the router talks to).

@openshift/networking

--- a/test/integration/router/router_http_server.go
+++ b/test/integration/router/router_http_server.go
@@ -9,6 +9,7 @@ import (
 	"net/http"
 	"os"
 	"strings"
+	"sync"
 
 	"github.com/golang/glog"
 	"golang.org/x/net/websocket"
@@ -111,6 +112,7 @@ type TestHttpService struct {
 	NamespaceListResponse string
 
 	listeners []net.Listener
+	wg        sync.WaitGroup
 }
 
 const (
@@ -290,21 +292,16 @@ func (s *TestHttpService) handleWebSocket(ws *websocket.Conn) {
 
 // Stop stops the service by closing any registered listeners
 func (s *TestHttpService) Stop() {
-	if s.listeners != nil && len(s.listeners) > 0 {
-		for _, l := range s.listeners {
-			if l != nil {
-				glog.Infof("Stopping listener at %s\n", l.Addr().String())
-				l.Close()
-			}
-		}
+	for _, l := range s.listeners {
+		glog.Infof("Stopping listener at %s\n", l.Addr().String())
+		l.Close()
 	}
+	s.wg.Wait()
 }
 
 // Start will start the http service to simulate the master and client urls.  It sets up the appropriate watch
 // endpoints and serves the secure and non-secure traffic.
 func (s *TestHttpService) Start() error {
-	s.listeners = make([]net.Listener, 3)
-
 	if err := s.startMaster(); err != nil {
 		return err
 	}
@@ -424,12 +421,13 @@ func (s *TestHttpService) startServing(addr string, handler http.Handler) error 
 
 	glog.Infof("Started, serving at %s\n", listener.Addr().String())
 
+	s.wg.Add(1)
 	go func() {
 		err := http.Serve(listener, handler)
-
-		if err != nil {
+		if err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
 			glog.Errorf("HTTP server failed: %v", err)
 		}
+		s.wg.Done()
 	}()
 
 	return nil
@@ -456,10 +454,13 @@ func (s *TestHttpService) startServingTLS(addr string, cert []byte, key []byte, 
 	s.listeners = append(s.listeners, listener)
 	glog.Infof("Started, serving TLS at %s\n", listener.Addr().String())
 
+	s.wg.Add(1)
 	go func() {
-		if err := http.Serve(listener, handler); err != nil {
+		err := http.Serve(listener, handler)
+		if err != nil && !strings.Contains(err.Error(), "use of closed network connection") {
 			glog.Errorf("HTTPS server failed: %v", err)
 		}
+		s.wg.Done()
 	}()
 
 	return nil

